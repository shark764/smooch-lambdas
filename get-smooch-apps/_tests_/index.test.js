const { validateTenantPermissions, validatePlatformPermissions } = require('serenova-js-utils/lambda/api');

jest.mock('aws-sdk');
jest.mock('serenova-js-utils/lambda/api');

validateTenantPermissions.mockReturnValue(true);
validatePlatformPermissions.mockReturnValue(true);

beforeAll(() => {
  global.process.env = {
    AWS_REGION: 'us-east-1',
    ENVIRONMENT: 'dev',
  };
});

const event = {
  params: {
    'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
  },
  identity: {
    'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
};

const mockQuery = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({
      Count: 1,
      Items: [{
        'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
      }],
    }),
  }));

jest.mock('aws-sdk', () => ({
  config: {
    update: jest.fn(),
  },
  DynamoDB: {
    DocumentClient: jest.fn().mockImplementation(() => ({ query: mockQuery })),
  },
}));

const { handler } = require('../index');

describe('get-smooch-apps', () => {
  describe('Everything is successful', () => {
    it('sends back status 200 if the code runs without error ', async () => {
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });
    describe('Walkthrough', () => {
      beforeAll(async () => {
        await handler(event);
      });
      it('passes in the correct arguments to validateTenantPermissions', async () => {
        expect(validateTenantPermissions.mock.calls).toMatchSnapshot();
      });
      it('passes in the correct arguments to validatePlatformPermissions', async () => {
        expect(validatePlatformPermissions.mock.calls).toMatchSnapshot();
      });
      it('passes in the correct arguments to docClient.query()', async () => {
        expect(mockQuery.mock.calls).toMatchSnapshot();
      });
    });
  });

  it('sends back status 400 when there are invalid params value', async () => {
    const mockEvent = {
      params: {
        'tenant-id': '',
      },
      identity: {
        'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
      },
    };
    const result = await handler(mockEvent);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 400 when there are not enough permissions', async () => {
    validateTenantPermissions.mockReturnValueOnce(false);
    validatePlatformPermissions.mockReturnValueOnce(false);
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a problem fetching apps in DynamoDB', async () => {
    mockQuery.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });
});
