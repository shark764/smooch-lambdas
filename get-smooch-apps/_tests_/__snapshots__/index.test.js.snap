// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-smooch-apps Everything is successful Walkthrough passes in the correct arguments to docClient.query() 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#integrationType": "type",
        "#tenantId": "tenant-id",
      },
      "ExpressionAttributeValues": Object {
        ":t": "66d83870-30df-4a3b-8801-59edff162034",
        ":type": "app",
      },
      "IndexName": "tenant-id-type-index",
      "KeyConditionExpression": "#tenantId = :t and #integrationType = :type",
      "TableName": "us-east-1-dev-smooch",
    },
  ],
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#integrationType": "type",
        "#tenantId": "tenant-id",
      },
      "ExpressionAttributeValues": Object {
        ":t": "66d83870-30df-4a3b-8801-59edff162034",
        ":type": "app",
      },
      "IndexName": "tenant-id-type-index",
      "KeyConditionExpression": "#tenantId = :t and #integrationType = :type",
      "TableName": "us-east-1-dev-smooch",
    },
  ],
]
`;

exports[`get-smooch-apps Everything is successful Walkthrough passes in the correct arguments to validatePlatformPermissions 1`] = `
Array [
  Array [
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "PLATFORM_DIGITAL_CHANNELS_APP",
    ],
  ],
  Array [
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "PLATFORM_DIGITAL_CHANNELS_APP",
    ],
  ],
]
`;

exports[`get-smooch-apps Everything is successful Walkthrough passes in the correct arguments to validateTenantPermissions 1`] = `
Array [
  Array [
    "66d83870-30df-4a3b-8801-59edff162034",
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "DIGITAL_CHANNELS_APP_READ",
    ],
  ],
  Array [
    "66d83870-30df-4a3b-8801-59edff162034",
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "DIGITAL_CHANNELS_APP_READ",
    ],
  ],
]
`;

exports[`get-smooch-apps Everything is successful sends back status 200 if the code runs without error  1`] = `
Object {
  "body": Object {
    "result": Array [
      Object {
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
    ],
  },
  "status": 200,
}
`;

exports[`get-smooch-apps sends back status 400 when there are invalid params value 1`] = `
Object {
  "body": Object {
    "error": [ValidationError: "tenant-id" is not allowed to be empty],
    "message": "Error: invalid params value \\"tenant-id\\" is not allowed to be empty",
  },
  "status": 400,
}
`;

exports[`get-smooch-apps sends back status 400 when there are not enough permissions 1`] = `
Object {
  "body": Object {
    "message": "Error not enough permissions",
  },
  "status": 400,
}
`;

exports[`get-smooch-apps sends back status 500 when there is a problem fetching apps in DynamoDB 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to fetch apps in DynamoDB",
  },
  "status": 500,
}
`;
